### PR TITLE
Metrics forwarder pko setup

### PR DESCRIPTION
**Pull request by Tafhim**
This PR resolves both [OSD-14716](https://issues.redhat.com/browse/OSD-14716) and [OSD-14718](https://issues.redhat.com/browse/OSD-14718). This is part of the Epic [SDE-1991](https://issues.redhat.com/browse/SDE-1991).

It adds package that is deployed on HCP Namespaces. It also add the policies that are going to be used to deploy CMO Config to hosted clusters and the package to the HCP namespaces.

### CODE REVIEW

**Comment by supreeth7** [comment]
Can you please add a description?

File: README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Metrics Forwarder
```
Comment (supreeth7) [inline README.md:1]:
```suggestion
# Dataplane Metrics Forwarder
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+The Metrics Forwarder is a Proxy that lives in each Hosted Control Plane (HCP) Namespace in Management Clusters. The main goal of this Proxy is to let the relevant Hosted Cluster's Data Plane push metrics to RHOBS. This proxy exposes a route that is known to the Cluster Monitoring Operator (CMO) running in the Data Plane. The CMO pushes the metrics to this route. And the proxy simply forwards this request to the Openshift Observability Operator (OBO) running on the relevant Management Cluster. This way the metrics ultimately end up in RHOBS.
```
Comment (supreeth7) [inline README.md:3]:
```suggestion
The dataplane metrics forwarder is a proxy that lives in each hosted control plane (HCP) namespace in management clusters. The main goal of this proxy is to let the hosted clusters push metrics to RHOBS. This proxy exposes a route that is known to the Cluster Monitoring Operator (CMO) running in the dataplane. The CMO pushes the metrics to this route. And the proxy simply forwards this request to the OpenShift Observability Operator (OBO) running on the relevant management cluster. This way the dataplane metrics ultimately end up in RHOBS.
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+## Deployment process
+
+The proxy is deployed using ACM policy. The ACM Policy targets all Management Clusters and only HCP Namespaces in those Management Clusters. It deploys a PKO Package which deploys the resources the proxy needs. The resources it deploys are:
```
Comment (supreeth7) [inline README.md:7]:
```suggestion
The proxy is deployed using ACM policy. The ACM policy targets all management clusters and only HCP namespaces in those management clusters. It deploys a PKO package which deploys the resources the proxy needs. The resources it deploys are:
```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- RoleBinding
+- CronJob
+- Issuer
+- Certificate
+- ConfigMap
+- Deployment
+- Service
+- Route
+
+Additionally, the CMO on the Data Plane is configured using another Policy that only targets Hosted Clusters.
```
Comment (supreeth7) [inline README.md:19]:
```suggestion
Additionally, the CMO on the dataplane is configured using another policy that only targets hosted clusters.
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+## Proxy initialization process
+
+The Proxy needs to be initialized to work properly. This phase might not be needed later if the root-ca contains certificates that are compatible with Cert Manager.
+Initially, PKO Package deploys 4 resources:
```
Comment (supreeth7) [inline README.md:24]:
```suggestion
The proxy needs to be initialized to work properly. This phase might not be needed later if the root-ca contains certificates that are compatible with [cert-manager](https://cert-manager.io/).
Initially, PKO package deploys 4 resources:
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+- ServiceAccount
+- Role
+- RoleBinding
+- CronJob
+
+This CronJob mianly recreates the root-ca in the HCP with names that are compatible with Cert Manager.
```
Comment (supreeth7) [inline README.md:30]:
```suggestion
This CronJob mainly recreates the root-ca in the HCP with names that are compatible with cert-manager. 
```

File: dataplane-policy.yaml [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+                    metadata:
+                      name: cluster-monitoring-config
+                      namespace: openshift-monitoring
+                    data:
+                      config.yaml: |
+                        enableUserWorkload: true
+                        prometheusK8s:
+                        # If the location of this config changes,
+                        # make sure to adjust https://github.com/openshift/ops-sop/blob/master/v4/howto/observability/o11m_mst_killswitch.md
+                        # so that killswitch instructions point to the right file!
```
Comment (supreeth7) [inline dataplane-policy.yaml:37]:
```suggestion
```
This can be removed

File: dataplane-policy.yaml [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        kind: ConfigurationPolicy
+        metadata:
+            name: metrics-forwarder-config
+        spec:
+            remediationAction: enforce
+            evaluationInterval:
+                compliant: 2h
+                noncompliant: 45s
+            object-templates:
+                - complianceType: musthave
```
Comment (supreeth7) [inline dataplane-policy.yaml:24]:
```suggestion
                - complianceType: mustonlyhave
```
This is related to a bug, more details in https://issues.redhat.com/browse/OSD-15177

File: dataplane-policy.yaml [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                              - sourceLabels: [__name__]
+                                action: keep
+                              queueConfig:
+                                capacity: 2500
+                                maxShards: 1000
+                                minShards: 1
+                                maxSamplesPerSend: 500
+                                batchSendDeadline: 60s
+                                minBackoff: 30ms
+                                maxBackoff: 5s
```
Comment (supreeth7) [inline dataplane-policy.yaml:57]:
```suggestion
                                capacity: 10000
                                maxShards: 500
                                minShards: 1
                                maxSamplesPerSend: 2000
                                batchSendDeadline: 60s
                                minBackoff: 30ms
                                maxBackoff: 30s
```

File: dataplane-policy.yaml [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+                            - effect: NoSchedule
+                              key: node-role.kubernetes.io/worker
+                              operator: Exists
+                        grafana:
+                          nodeSelector:
+                            node-role.kubernetes.io/worker: ""
+                          tolerations:
+                            - effect: NoSchedule
+                              key: node-role.kubernetes.io/worker
+                              operator: Exists
```
Comment (supreeth7) [inline dataplane-policy.yaml:110]:
```suggestion
```
We do not need this, ref: https://issues.redhat.com/browse/OSD-14556

File: package/manifest.yaml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  availabilityProbes:
+    - selector:
+        kind:
+          groups: cert-manager.io
+          kind: Issuer
+      probes:
+      - condition:
+          type: Ready
+          status: "True"
+
```
Comment (supreeth7) [inline package/manifest.yaml:42]: No new line EOF

**Review by supreeth7 — CHANGES_REQUESTED** [review]
Overall looks good, thanks for all the hard work!
A few minor issues to be addressed.

File: README.md (hunk 4 above)
Comment (wanghaoran1988) [reply README.md:19]: policy -> ACM policy?

File: README.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+The Proxy needs to be initialized to work properly. This phase might not be needed later if the root-ca contains certificates that are compatible with Cert Manager.
+Initially, PKO Package deploys 4 resources:
+- ServiceAccount
+- Role
+- RoleBinding
+- CronJob
+
+This CronJob mianly recreates the root-ca in the HCP with names that are compatible with Cert Manager. 
+
+## Contact
```
Comment (wanghaoran1988) [inline README.md:32]: Let's remove this part? since the channel are redhat internal slack channel.

File: dataplane-policy.yaml (hunk 7 above)
Comment (wanghaoran1988) [reply dataplane-policy.yaml:37]: +1

File: dataplane-policy.yaml [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+                              queueConfig:
+                                capacity: 2500
+                                maxShards: 1000
+                                minShards: 1
+                                maxSamplesPerSend: 500
+                                batchSendDeadline: 60s
+                                minBackoff: 30ms
+                                maxBackoff: 5s
+                              tlsConfig:
+                                insecureSkipVerify: true
```
Comment (wanghaoran1988) [inline dataplane-policy.yaml:59]: Does this has to be true?

File: package/resources.yaml.gotmpl [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+kind: Certificate
+metadata:
+  annotations:
+    package-operator.run/phase: certman-cert
+  name: metrics-forwarder-cert
+spec:
+  # Secret names are always required.
+  secretName: metrics-forwarder-secret
+
+  # secretTemplate is optional. If set, these annotations and labels will be
```
Comment (wanghaoran1988) [inline package/resources.yaml.gotmpl:115]: remove the comment here?

File: package/resources.yaml.gotmpl [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+  # copied to the Secret named example-com-tls. These labels and annotations will
+  # be re-reconciled if the Certificate's secretTemplate changes. secretTemplate
+  # is also enforced, so relevant label and annotation changes on the Secret by a
+  # third party will be overwriten by cert-manager to match the secretTemplate.
+  secretTemplate:
+    labels:
+      app: metrics-forwarder
+
+  duration: 21600h
```
Comment (wanghaoran1988) [inline package/resources.yaml.gotmpl:124]: This is 30 month, let's make it 25960, so it's 36 month?

File: dataplane-policy.yaml (hunk 13 above)
Comment (supreeth7) [reply dataplane-policy.yaml:56]: Had the same doubt, it is `true` in our default CMO config.

File: package/resources.yaml.gotmpl [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+      app: metrics-forwarder
+
+  duration: 21600h 
+  renewBefore: 360h # 15d
+  subject:
+    organizations:
+      - openshift
+  # The use of the common name field has been deprecated since 2000 and is
+  # discouraged from being used.
+  commonName:
```
Comment (wanghaoran1988) [inline package/resources.yaml.gotmpl:131]: Let's remove this? since it's deprecated since 2000 in the comment.

File: README.md (hunk 12 above)
Comment (wanghaoran1988) [reply README.md:32]: We can ignore that now.

File: dataplane-policy.yaml (hunk 13 above)
Comment (wanghaoran1988) [reply dataplane-policy.yaml:56]: should be, but worth a test to see whether it works or not.

**Pull request merged**